### DISCUSSION TITLE
Low GPU utilization

### DISCUSSION THREAD

**Discussion by jbuckman** [Lightning-AI/pytorch-lightning#20292] [DDP / multi-GPU / multi-node]
This is my first time trying out PyTorch Lightning. On a node with 8 H100s, I'm running the code from this example:  https://lightning.ai/lightning-ai/studios/pretrain-an-llm-with-pytorch-lightning?tab=overview

All three versions, small medium large, get extremely low GPU utilization, nvidia-smi fluctuating but typically around 50%. Is the example code not correct? What changes need to be made?

**Comment by linyanxiang26** [comment]
I'm experiencing almost same issue. After completing a 50,000-step training run with two A40 GPUs, I observed very high and severe fluctuations. The only difference is that my fluctuated GPU utilization peaked at 100%. I'm suspecting this is due to a mismatch in data throughput between the CPU pre-processing and GPU processing. Have you found any effective solutions? Thanks in advance.

**Comment by arodland** [comment]
Try increasing `num_workers` on the dataloader. If 4 gets you 50% utilization, try 8.

In general, look at CPU and I/O utilization on the system to try to find out where the bottleneck is.

**Comment by iamgooold** [comment]
Hey @jbuckman @linyanxiang26, 50% GPU util on H100s is definitely a data/CPU bottleneck, not the model.

**Most common fixes, in order of impact:**

**1. Dataloader workers** — like @arodland said, but go higher on H100s:
```python
train_dataloader = DataLoader(
    dataset, 
    batch_size=batch_size,
    num_workers=16,  # Start here for H100. Rule: 4-8 per GPU
    pin_memory=True,
    persistent_workers=True,  # Keeps workers alive between epochs
    prefetch_factor=4  # Default is 2
)

**Comment by abhinavsriva** [comment]
Hey @jbuckman and @linyanxiang26, I was hitting this exact same low utilization bottlenecks.

Instead of guessing, I built an open-source callback called TraceML that plugs directly into your Lightning Trainer. It gives you a visual diagnostic dashboard that explicitly shows if your run is INPUT-BOUND, COMPUTE-BOUND, or WAIT-HEAVY.

It’s a one-line add to your callbacks list:

```
from traceml.integrations.lightning import TraceMLCallback
trainer = L.Trainer(callbacks=[TraceMLCallback()], ...)
```

You can find the docs here:  https://github.com/traceopt-ai/traceml/blob/main/docs/user_guide/integrations/lightning.md

Hope it helps
